### PR TITLE
[CPU] Avoid inserting additional transpose + reorder after RNN node.

### PR DESCRIPTION
**Pull request by nshchego**
### Details:
 - *Reorders before and after RNN Sequences were removed for case batch == 1*
 - *nGraph transformation reshape-transpose-reshape after RNN Sequences was replaced by plugin's optimization*

### Tickets:
 - *55822*

### CODE REVIEW

**Comment by dmitry-gorokhov** [comment]
@mandrono please review the PR

File: inference-engine/src/mkldnn_plugin/mkldnn_graph_optimizer.cpp [hunk 1]
```diff
@@ -1812,3 +1817,57 @@ void MKLDNNGraphOptimizer::MergeTransposeAndReorder(MKLDNNGraph &graph) {
...
+    auto parent = graphNodes.begin();
+    while (parent != graphNodes.end()) {
+        auto parentNode = *parent;
+        if (!isSutableParentNode(parentNode)) {
+            parent++;
+            continue;
+        }
+
+        auto childNode = parentNode->getChildEdgeAt(0)->getChild();
+        if (!isSutableChildNode(parentNode, childNode)) {
```
Comment (mandrono) [inline inference-engine/src/mkldnn_plugin/mkldnn_graph_optimizer.cpp:1842]: not sure that we really need this check
but if we did it we need to validate all child for 0 port edges

File: inference-engine/src/mkldnn_plugin/mkldnn_graph_optimizer.cpp [hunk 2]
```diff
@@ -1812,3 +1817,57 @@ void MKLDNNGraphOptimizer::MergeTransposeAndReorder(MKLDNNGraph &graph) {
...
+            parent++;
+            continue;
+        }
+
+        auto parents = childNode->parentEdges;
```
Comment (mandrono) [inline inference-engine/src/mkldnn_plugin/mkldnn_graph_optimizer.cpp:1847]: i believe better iterate through ```getChildEdgesAtPort(0)``` to account for the case with multiple edges

File: inference-engine/src/mkldnn_plugin/mkldnn_graph_optimizer.cpp [hunk 3]
```diff
@@ -1812,3 +1817,57 @@ void MKLDNNGraphOptimizer::MergeTransposeAndReorder(MKLDNNGraph &graph) {
...
+            auto parent = p_edge->getParent();
+            if (!parent) continue;
+
+            if (parent == parentNode) {
+                auto newRnnOutDims = parentNode->outDims[0].ToSizeVector();
+                newRnnOutDims.erase(newRnnOutDims.begin() + 1);
+                parentNode->outDims[0] = MKLDNNDims{newRnnOutDims};
+
+                const MKLDNNNodePtr newReshape = std::make_shared<MKLDNNReshapeNode>(parent->getName() + "_abc_a1bc", graph.getEngine(), graph.weightsCache);
+                newReshape->inDims.push_back(parentNode->outDims[0]);
```
Comment (mandrono) [inline inference-engine/src/mkldnn_plugin/mkldnn_graph_optimizer.cpp:1861]: can we fill inDims/outDims and precision into ctor?

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_rnn.cpp [hunk 4]
```diff
@@ -406,6 +406,9 @@ void MKLDNNRNN::fillSeqDesc() {
 
     if (nativeOrder)
         in_candidate.push_back(MKLDNNMemoryDesc{inDims[RNNInOutKind::Layer], dataType, memory::format_tag::tnc});
+    else if (N == 1 && (inputMemoryFormatsFilter.empty() || inputMemoryFormatsFilter[0] == memory::format_tag::tnc))
```
[the remark this answers is not in the document]
Comment (nshchego) [reply inference-engine/src/mkldnn_plugin/nodes/mkldnn_rnn.cpp:409]: inputMemoryFormatsFilter is used to filter primitive descriptors and filled from nGraph runtime info. It doesn't matter how this runtime info was created, we need to check it here.

File: inference-engine/src/mkldnn_plugin/mkldnn_graph_optimizer.cpp (hunk 1 above)
Comment (nshchego) [reply inference-engine/src/mkldnn_plugin/mkldnn_graph_optimizer.cpp]: Removed

File: inference-engine/src/mkldnn_plugin/mkldnn_graph_optimizer.cpp (hunk 2 above)
Comment (nshchego) [reply inference-engine/src/mkldnn_plugin/mkldnn_graph_optimizer.cpp]: Done

File: inference-engine/src/mkldnn_plugin/mkldnn_graph_optimizer.cpp (hunk 3 above)
Comment (nshchego) [reply inference-engine/src/mkldnn_plugin/mkldnn_graph_optimizer.cpp]: Fixed: parent node is passed into ctor and dims with precisions are filled there. Child node was not due to in general it may not exist when this ctor is called.

**Comment by dmitry-gorokhov** [comment]
@nshchego @mandrono is there anything extra to be done here or the PR is pending my review?

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_rnn.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (mandrono) [reply inference-engine/src/mkldnn_plugin/nodes/mkldnn_rnn.cpp:409]: I agree, but this attribute used only for test purpose
we have special method to filter primitive descriptors - ```filterSupportedPrimitiveDescriptors```
so I guess we shouldn't mix test part and execution part

File: inference-engine/src/mkldnn_plugin/mkldnn_graph_optimizer.cpp [hunk 5]
```diff
@@ -1802,3 +1807,42 @@ void MKLDNNGraphOptimizer::MergeTransposeAndReorder(MKLDNNGraph &graph) {
...
+        newRnnOutDims.erase(newRnnOutDims.begin() + 1);
+        parentNode->outDims[0] = MKLDNNDims{newRnnOutDims};
+
+        for (size_t i = 0; i < childrenEdges.size(); i++) {
+            auto edge = childrenEdges[i];
+            auto childNode = edge->getChild();
+
+            const MKLDNNNodePtr newReshape = std::make_shared<MKLDNNReshapeNode>(parentNode,
+                    parentNode->getName() + "_abc_a1bc_" + std::to_string(i), graph.getEngine(), graph.weightsCache);
+            newReshape->outDims.push_back(childNode->inDims[edge->getOutputNum()]);
```
Comment (mandrono) [inline inference-engine/src/mkldnn_plugin/mkldnn_graph_optimizer.cpp:1840]: I believe better to pass inDims, outDims and precision directly in ctor

**Review by mandrono — COMMENTED** [review]
i guess also need to update tests to check number of reorders after RNN node

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_rnn.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (nshchego) [reply inference-engine/src/mkldnn_plugin/nodes/mkldnn_rnn.cpp:409]: This "test part" is already in the base class ctor - MKLDNNNode and such tests requires this condition. So we need this check here or delete these tests.

[the remark this answers is not in the document]
Comment (mandrono) [reply inference-engine/src/mkldnn_plugin/nodes/mkldnn_rnn.cpp:409]: for which test is required this check?

[the remark this answers is not in the document]
Comment (nshchego) [reply inference-engine/src/mkldnn_plugin/nodes/mkldnn_rnn.cpp:409]: Some tests which set Fmt.

File: inference-engine/src/mkldnn_plugin/mkldnn_graph_optimizer.cpp (hunk 5 above)
Comment (nshchego) [reply inference-engine/src/mkldnn_plugin/mkldnn_graph_optimizer.cpp]: FIxed

**Review by mandrono — APPROVED** [review]

**Comment by dmitry-gorokhov** [comment]
@nshchego please assign on me once CI will be fixed

**Comment by nshchego** [comment]
@dmitry-gorokhov please take a look

File: inference-engine/src/mkldnn_plugin/mkldnn_graph.cpp [hunk 6]
```diff
@@ -515,7 +515,7 @@ static inline bool isConstOutput(MKLDNNEdgePtr edge) {
     return edge->getParent()->isConstant() && !edge->getChild()->isConstant();
 }
 
-static edge_clusters_t findEdgeClusters(const std::vector<MKLDNNEdgePtr> & graphEdges) {
+static edge_clusters_t findEdgeClusters(const std::deque<MKLDNNEdgePtr> & graphEdges) {
```
Comment (dmitry-gorokhov) [inline inference-engine/src/mkldnn_plugin/mkldnn_graph.cpp:518]: Lets move deque related changes into separate PR.

**Review by dmitry-gorokhov — APPROVED** [review]
LGTM. Can be merged after internal validation.

File: inference-engine/src/mkldnn_plugin/mkldnn_graph.cpp (hunk 6 above)
Comment (nshchego) [reply inference-engine/src/mkldnn_plugin/mkldnn_graph.cpp]: Done
